File: py/register_orderly_account.py
"""
Registers an account on Orderly Network using a Privy agentic wallet
"""
import os
import sys
import argparse
import time
import base64
from dotenv import load_dotenv
import requests

load_dotenv()

# Configuration
ORDERLY_API_URL = "https://api.orderly.org"  # or https://testnet-api.orderly.org for testnet
BROKER_ID = "woofi_pro"
VERIFYING_CONTRACT = "0xCcCCccccCCCCcCCCCCCcCcCccCcCCCcCcccccccC"

# Privy API base URL
PRIVY_API_BASE = "https://auth.privy.io/api/v1"


def get_wallet_address(wallet_id: str, app_id: str, app_secret: str) -> str:
    """Get wallet address from Privy"""
    auth_string = f"{app_id}:{app_secret}"
    encoded_auth = base64.b64encode(auth_string.encode()).decode()
    headers = {
        "Authorization": f"Basic {encoded_auth}",
        "privy-app-id": app_id,
        "Content-Type": "application/json",
    }
    
    response = requests.get(
        f"{PRIVY_API_BASE}/wallets/{wallet_id}",
        headers=headers
    )
    
    if not response.ok:
        raise Exception(f"Failed to get wallet: {response.text}")
    
    wallet = response.json()
    wallet_address = (
        wallet.get("address") or
        (wallet.get("addresses", [{}])[0].get("address") if wallet.get("addresses") else None) or
        (wallet.get("addresses", [None])[0] if wallet.get("addresses") else None)
    )
    
    if not wallet_address:
        raise Exception("Could not determine wallet address from wallet object")
    
    return wallet_address


def sign_typed_data(wallet_id: str, typed_data: dict, app_id: str, app_secret: str, authorization_secret: str) -> str:
    """Sign EIP-712 typed data using Privy"""
    auth_string = f"{app_id}:{app_secret}"
    encoded_auth = base64.b64encode(auth_string.encode()).decode()
    headers = {
        "Authorization": f"Basic {encoded_auth}",
        "privy-app-id": app_id,
        "Content-Type": "application/json",
    }
    
    payload = {
        "authorization_context": {
            "authorization_private_keys": [authorization_secret]
        },
        "params": {
            "typed_data": typed_data
        }
    }
    
    response = requests.post(
        f"{PRIVY_API_BASE}/wallets/{wallet_id}/ethereum/sign-typed-data",
        headers=headers,
        json=payload
    )
    
    if not response.ok:
        raise Exception(f"Failed to sign typed data: {response.text}")
    
    result = response.json()
    return result.get("signature")


def register_orderly_account(wallet_id: str, wallet_address: str = None, chain_id: str = "421614") -> dict:
    """
    Registers an account on Orderly Network using a Privy agentic wallet
    
    Args:
        wallet_id: The wallet ID to use for registration
        wallet_address: The wallet address (optional, will be fetched if not provided)
        chain_id: Chain ID (optional, defaults to 421614 for Arbitrum Sepolia)
        
    Returns:
        The registration response
    """
    # Validate environment variables
    app_id = os.getenv("PRIVY_APP_ID")
    app_secret = os.getenv("PRIVY_APP_SECRET")
    authorization_id = os.getenv("PRIVY_AUTHORIZATION_ID")
    authorization_secret = os.getenv("PRIVY_AUTHORIZATION_SECRET")
    
    if not app_id or not app_secret:
        raise ValueError(
            "Missing required environment variables. Please ensure the following are set in your .env file:\n"
            "- PRIVY_APP_ID\n"
            "- PRIVY_APP_SECRET"
        )
    
    if not authorization_id or not authorization_secret:
        raise ValueError(
            "Missing authorization credentials. Please ensure the following are set in your .env file:\n"
            "- PRIVY_AUTHORIZATION_ID\n"
            "- PRIVY_AUTHORIZATION_SECRET"
        )
    
    if not wallet_id:
        raise ValueError("Wallet ID is required. Use --wallet-id <wallet_id>")
    
    try:
        # Get wallet address if not provided
        if not wallet_address:
            print("Fetching wallet details...")
            wallet_address = get_wallet_address(wallet_id, app_id, app_secret)
            print(f"   Wallet Address: {wallet_address}")
        
        chain_id_number = int(chain_id)
        chain_id_hex = f"0x{chain_id_number:x}"
        
        print("\nPreparing Orderly account registration...")
        print(f"   Wallet ID: {wallet_id}")
        print(f"   Wallet Address: {wallet_address}")
        print(f"   Chain ID: {chain_id} ({chain_id_hex})")
        print(f"   Broker ID: {BROKER_ID}")
        print(f"   Orderly API: {ORDERLY_API_URL}")
        
        # Step 1: Get registration nonce from Orderly
        print("\nStep 1: Getting registration nonce...")
        nonce_response = requests.get(f"{ORDERLY_API_URL}/v1/registration_nonce")
        if not nonce_response.ok:
            raise Exception(f"Failed to get registration nonce ({nonce_response.status_code}): {nonce_response.text}")
        
        nonce_data = nonce_response.json()
        registration_nonce = nonce_data["data"]["registration_nonce"]
        print(f"   Registration Nonce: {registration_nonce}")
        
        # Step 2: Prepare EIP-712 typed data for Orderly registration
        timestamp = int(time.time() * 1000)  # UNIX milliseconds
        register_message = {
            "brokerId": BROKER_ID,
            "chainId": chain_id_number,
            "timestamp": timestamp,
            "registrationNonce": registration_nonce,
        }
        
        # EIP-712 types structure
        typed_data = {
            "domain": {
                "name": "Orderly",
                "version": "1",
                "chainId": chain_id_hex,
                "verifyingContract": VERIFYING_CONTRACT,
            },
            "message": register_message,
            "primary_type": "Registration",
            "types": {
                "EIP712Domain": [
                    {"name": "name", "type": "string"},
                    {"name": "version", "type": "string"},
                    {"name": "chainId", "type": "uint256"},
                    {"name": "verifyingContract", "type": "address"},
                ],
                "Registration": [
                    {"name": "brokerId", "type": "string"},
                    {"name": "chainId", "type": "uint256"},
                    {"name": "timestamp", "type": "uint64"},
                    {"name": "registrationNonce", "type": "uint256"},
                ],
            },
        }
        
        print("\nStep 2: Signing EIP-712 message...")
        print(f"Register message: {register_message}")
        
        # Step 3: Sign the typed data using Privy
        signature = sign_typed_data(wallet_id, typed_data, app_id, app_secret, authorization_secret)
        print(f"\n✅ Signature generated: {signature}")
        
        # Step 4: Register with Orderly API
        print("\nStep 3: Registering account with Orderly...")
        registration_payload = {
            "message": register_message,
            "signature": signature,
            "userAddress": wallet_address,
        }
        
        print(f"Registration payload: {registration_payload}")
        
        response = requests.post(
            f"{ORDERLY_API_URL}/v1/register_account",
            headers={"Content-Type": "application/json"},
            json=registration_payload
        )
        
        if not response.ok:
            raise Exception(f"Orderly API error ({response.status_code}): {response.text}")
        
        result = response.json()
        
        if not result.get("success"):
            raise Exception(f"Orderly registration failed: {result.get('message', 'Unknown error')}")
        
        orderly_account_id = result["data"]["account_id"]
        print("\n✅ Account registered successfully with Orderly!")
        print(f"Registration response: {result}")
        print(f"\n📝 Orderly Account ID: {orderly_account_id}")
        
        return {
            "signature": signature,
            "walletAddress": wallet_address,
            "orderlyAccountId": orderly_account_id,
            "registration": result,
        }
    except Exception as error:
        print(f"\n❌ Failed to register Orderly account: {error}")
        raise


def main():
    """Main execution function"""
    parser = argparse.ArgumentParser(description="Register an account on Orderly Network")
    parser.add_argument("--wallet-id", required=True, help="Wallet ID to use for registration")
    parser.add_argument("--wallet-address", help="Wallet address (optional, will be fetched if not provided)")
    parser.add_argument("--chain-id", default="421614", help="Chain ID (optional, default: 421614 = Arbitrum Sepolia)")
    
    args = parser.parse_args()
    
    try:
        result = register_orderly_account(
            wallet_id=args.wallet_id,
            wallet_address=args.wallet_address,
            chain_id=args.chain_id
        )
        
        print("\n📝 Registration Summary:")
        print(f"   Wallet Address: {result['walletAddress']}")
        print(f"   Orderly Account ID: {result.get('orderlyAccountId', 'N/A')}")
        
        sys.exit(0)
    except Exception as error:
        print(f"\n❌ Failed to register Orderly account: {error}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: js/registerOrderlyAccount.js
require("dotenv").config();
const { PrivyClient } = require("@privy-io/node");

/**
 * Registers an account on Orderly Network using a Privy agentic wallet
 *
 * @param {Object} options - Configuration options
 * @param {string} options.walletId - The wallet ID to use for registration
 * @param {string} options.walletAddress - The wallet address (optional, will be fetched if not provided)
 * @param {string} options.chainId - Chain ID (optional, defaults to 421614 for Arbitrum Sepolia)
 * @returns {Promise<Object>} The registration response
 */
async function registerOrderlyAccount(options = {}) {
  // Validate environment variables
  const appId = process.env.PRIVY_APP_ID;
  const appSecret = process.env.PRIVY_APP_SECRET;
  const authorizationId = process.env.PRIVY_AUTHORIZATION_ID;
  const authorizationSecret = process.env.PRIVY_AUTHORIZATION_SECRET;

  if (!appId || !appSecret) {
    throw new Error(
      "Missing required environment variables. Please ensure the following are set in your .env file:\n" +
        "- PRIVY_APP_ID\n" +
        "- PRIVY_APP_SECRET"
    );
  }

  if (!authorizationId || !authorizationSecret) {
    throw new Error(
      "Missing authorization credentials. Please ensure the following are set in your .env file:\n" +
        "- PRIVY_AUTHORIZATION_ID\n" +
        "- PRIVY_AUTHORIZATION_SECRET"
    );
  }

  if (!options.walletId) {
    throw new Error("Wallet ID is required. Use --wallet-id <wallet_id>");
  }

  // Initialize Privy client
  const privy = new PrivyClient({
    appId: appId,
    appSecret: appSecret,
  });

  // Default to Arbitrum Sepolia testnet (chain ID: 421614)
  // Orderly supports multiple chains - check their docs for mainnet chain IDs
  const chainId = options.chainId || "421614"; // Arbitrum Sepolia
  const chainIdNumber = parseInt(chainId, 10);
  const chainIdHex = `0x${BigInt(chainId).toString(16)}`;

  // Orderly API endpoints
  // Testnet: https://testnet-api.orderly.org
  // Mainnet: https://api.orderly.org
  const orderlyApiUrl = "https://api.orderly.org";

  // Broker ID - default to "woofi_dex", can be customized
  const brokerId = "woofi_pro";

  // Create authorization context with the authorization private key
  const authorizationContext = {
    authorization_private_keys: [authorizationSecret],
  };

  try {
    // Get wallet address if not provided
    let walletAddress = options.walletAddress;
    if (!walletAddress) {
      console.log("Fetching wallet details...");
      // Use the get() method to retrieve wallet by ID
      // Reference: https://docs.privy.io/wallets/wallets/get-a-wallet/get-wallet-by-id
      const wallet = await privy.wallets().get(options.walletId);

      if (!wallet) {
        throw new Error(`Wallet with ID ${options.walletId} not found`);
      }

      // Extract address from wallet object
      walletAddress =
        wallet.address ||
        wallet.addresses?.[0]?.address ||
        wallet.addresses?.[0];

      if (!walletAddress) {
        throw new Error(
          "Could not determine wallet address from wallet object"
        );
      }
      console.log(`   Wallet Address: ${walletAddress}`);
    }

    console.log("\nPreparing Orderly account registration...");
    console.log(`   Wallet ID: ${options.walletId}`);
    console.log(`   Wallet Address: ${walletAddress}`);
    console.log(`   Chain ID: ${chainId} (${chainIdHex})`);
    console.log(`   Broker ID: ${brokerId}`);
    console.log(`   Orderly API: ${orderlyApiUrl}`);

    // Step 1: Get registration nonce from Orderly
    // Reference: https://orderly.network/docs/build-on-omnichain/user-flows/accounts
    console.log("\nStep 1: Getting registration nonce...");
    const nonceResponse = await fetch(`${orderlyApiUrl}/v1/registration_nonce`);
    if (!nonceResponse.ok) {
      const errorText = await nonceResponse.text();
      throw new Error(
        `Failed to get registration nonce (${nonceResponse.status}): ${errorText}`
      );
    }
    const nonceData = await nonceResponse.json();
    const registrationNonce = nonceData.data.registration_nonce;
    console.log(`   Registration Nonce: ${registrationNonce}`);

    // Step 2: Prepare EIP-712 typed data for Orderly registration
    // Based on Orderly documentation: https://orderly.network/docs/build-on-omnichain/user-flows/accounts
    const domain = {
      name: "Orderly",
      version: "1",
      chainId: chainIdHex, // Hex format for EIP-712
      verifyingContract: "0xCcCCccccCCCCcCCCCCCcCcCccCcCCCcCcccccccC",
    };

    // Message structure according to Orderly documentation
    const timestamp = Date.now(); // UNIX milliseconds
    const registerMessage = {
      brokerId: brokerId,
      chainId: chainIdNumber, // Number format for the message
      timestamp: timestamp,
      registrationNonce: registrationNonce,
    };

    // EIP-712 types structure
    const typedData = {
      domain: domain,
      message: registerMessage,
      primary_type: "Registration", // snake_case as required by Privy
      types: {
        EIP712Domain: [
          { name: "name", type: "string" },
          { name: "version", type: "string" },
          { name: "chainId", type: "uint256" },
          { name: "verifyingContract", type: "address" },
        ],
        Registration: [
          { name: "brokerId", type: "string" },
          { name: "chainId", type: "uint256" },
          { name: "timestamp", type: "uint64" },
          { name: "registrationNonce", type: "uint256" },
        ],
      },
    };

    console.log("\nStep 2: Signing EIP-712 message...");
    console.log("Register message:", JSON.stringify(registerMessage, null, 2));

    // Step 3: Sign the typed data using Privy
    const signatureResponse = await privy
      .wallets()
      .ethereum()
      .signTypedData(options.walletId, {
        authorization_context: authorizationContext,
        params: {
          typed_data: typedData,
        },
      });

    const signature = signatureResponse.signature;
    console.log(`\n✅ Signature generated: ${signature}`);

    // Step 4: Register with Orderly API
    // Endpoint: /v1/register_account
    // Reference: https://orderly.network/docs/build-on-omnichain/user-flows/accounts
    console.log("\nStep 3: Registering account with Orderly...");
    const registrationPayload = {
      message: registerMessage,
      signature: signature,
      userAddress: walletAddress,
    };

    console.log(
      "Registration payload:",
      JSON.stringify(registrationPayload, null, 2)
    );

    const response = await fetch(`${orderlyApiUrl}/v1/register_account`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(registrationPayload),
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(`Orderly API error (${response.status}): ${errorText}`);
    }

    const result = await response.json();

    if (!result.success) {
      throw new Error(
        `Orderly registration failed: ${result.message || "Unknown error"}`
      );
    }

    const orderlyAccountId = result.data.account_id;
    console.log("\n✅ Account registered successfully with Orderly!");
    console.log("Registration response:", JSON.stringify(result, null, 2));
    console.log(`\n📝 Orderly Account ID: ${orderlyAccountId}`);

    return {
      signature,
      walletAddress,
      orderlyAccountId,
      registration: result,
    };
  } catch (error) {
    console.error("\n❌ Failed to register Orderly account");
    if (error.message) {
      console.error("Error:", error.message);
    } else {
      console.error(error);
    }
    throw error;
  }
}

/**
 * Main execution function
 */
async function main() {
  try {
    // Parse command line arguments
    const args = process.argv.slice(2);
    const options = {};

    for (let i = 0; i < args.length; i++) {
      if (args[i] === "--wallet-id" && args[i + 1]) {
        options.walletId = args[i + 1];
        i++;
      } else if (args[i] === "--wallet-address" && args[i + 1]) {
        options.walletAddress = args[i + 1];
        i++;
      } else if (args[i] === "--chain-id" && args[i + 1]) {
        options.chainId = args[i + 1];
        i++;
      } else if (args[i] === "--help" || args[i] === "-h") {
        console.log(`
Usage: node registerOrderlyAccount.js [options]

Options:
  --wallet-id <id>        Wallet ID to use for registration (required)
  --wallet-address <addr>  Wallet address (optional, will be fetched if not provided)
  --chain-id <id>         Chain ID (optional, default: 421614 = Arbitrum Sepolia)
  --help, -h              Show this help message

Examples:
  # Register account on Orderly
  node registerOrderlyAccount.js --wallet-id wal_xxx

  # Register with specific chain
  node registerOrderlyAccount.js --wallet-id wal_xxx --chain-id 421614

Supported Chains:
  - Arbitrum Sepolia: 421614 (testnet, recommended for testing)
  - Arbitrum One: 42161 (mainnet)
  - Other chains supported by Orderly (check their documentation)

Note: 
  - Make sure your wallet has been created using createAgenticWallet.js
  - The script will sign an EIP-712 message for account registration
  - Check Orderly's API documentation for the exact endpoint and payload structure
        `);
        process.exit(0);
      }
    }

    const result = await registerOrderlyAccount(options);

    console.log("\n📝 Registration Summary:");
    console.log(`   Wallet Address: ${result.walletAddress}`);
    console.log(`   Orderly Account ID: ${result.orderlyAccountId || "N/A"}`);

    process.exit(0);
  } catch (error) {
    console.error("\n❌ Failed to register Orderly account");
    process.exit(1);
  }
}

// Run if executed directly
if (require.main === module) {
  main();
}

module.exports = { registerOrderlyAccount };
